2 fixes for secure_services sample.

oyvindronningstad opened the pull request:
1. The sample uses CONFIG_SPM_SERVICE_* configs to toggle features in the app, but these configs only make sense in the SPM, and cannot be enabled in the app.
2. The prevalidate call in the sample currently returns false when it should return true.

hakonfam commented on samples/nrf9160/secure_services/Kconfig:
```diff
@@ -0,0 +1,25 @@
+menu "Secure Service Sample"
+
+config SAMPLE_SERVICE_RNG
+	bool "Enable code that the spm_request_random_number()"
```
words

oyvindronningstad replied:
I accidentally the spm_request_random_number() :P

Fixed

hakonfam commented on samples/nrf9160/secure_services/Kconfig:
```diff
@@ -0,0 +1,25 @@
+	  This requires the service to be enabled in SPM, via SPM_SERVICE_RNG.
+
+config SAMPLE_SERVICE_PREVALIDATE
+	bool "Enable code that the spm_prevalidate_b1_upgrade()"
```
words

hakonfam commented on samples/nrf9160/secure_services/src/main.c:
```diff
@@ -12,7 +12,7 @@
 #include <fw_info.h>
 
-#if defined(CONFIG_BOOTLOADER_MCUBOOT) || defined(CONFIG_SPM_SERVICE_RNG)
+#if defined(CONFIG_BOOTLOADER_MCUBOOT) || defined(CONFIG_SAMPLE_SERVICE_RNG)
```
Should we have a overlay file which enabled the two new kconfig options you define? Both to get them compiled under CI and help those that want to test it.

oyvindronningstad replied:
We want something like

    CONFIG_BOOTLOADER_MCUBOOT=y
    CONFIG_SECURE_BOOT=y
    CONFIG_SAMPLE_SERVICE_PREVALIDATE=y
    spm_CONFIG_SPM_SERVICE_PREVALIDATE=y
    mcuboot_CONFIG_EXT_API_PROVIDE_EXT_API_ENABLED=y

in the conf file but using prefixes like this is invalid, do you know how to set configs in spm and mcuboot?

hakonfam replied:
You can set command line arguments in testcase.yml AFAIK, then you should be able to provide overlay files to spm by using the `spm` prefix, `-Dspm_OVERLAY_CONF=something.conf` not sure about the exact name of the define to use.

oyvindronningstad replied:
This didn't really work out for me, I tried -Dspm_CONF_FILE and -Dspm_OVERLAY_CONFIG. spm_CONF_FILE has to have a path from the SPM's CMakeLists, and spm_OVERLAY_CONFIG seems to be applied to the app, not the SPM. @tejlmand is this possible to achieve from the app's folder?

hakonfam replied:
We can try to figure this out later.

tejlmand replied:
Overlay files to child images does work.
https://github.com/nrfconnect/sdk-nrf/pull/2630#issuecomment-681021770

But overlay files should be given with absolute paths:
https://github.com/zephyrproject-rtos/zephyr/issues/23825#issuecomment-656005730

`CONF_FILE` does support relative path, but that solution has not been thought through fully, as it has some flaws that you can see based on the linked issue.

tejlmand replied:
You can also specify config options directly, without an overlay, like `cmake -DCONFIG_SAMPLE_SERVICE_PREVALIDATE=y ...` or `cmake -Dspm_CONFIG_SAMPLE_SERVICE_PREVALIDATE=y ...`.

Note that: `SAMPLE_SERVICE_RNG` defaults to `y`

oyvindronningstad replied:
I tried specifying them directly via extra_args

    extra_args: -DCONFIG_BOOTLOADER_MCUBOOT=y -DCONFIG_SECURE_BOOT=y -DCONFIG_SAMPLE_SERVICE_PREVALIDATE=y -Dspm_CONFIG_SPM_SERVICE_PREVALIDATE=y -Dmcuboot_CONFIG_EXT_API_PROVIDE_EXT_API_ENABLED=y

But I get this warning (when running through sanitycheck, not when doing it manually with west build)

    CMake Warning:
      Manually-specified variables were not used by the project:
    
        -DCONFIG_BOOTLOADER_MCUBOOT
        -DCONFIG_SAMPLE_SERVICE_PREVALIDATE
        -DCONFIG_SECURE_BOOT
        -Dmcuboot_CONFIG_EXT_API_PROVIDE_EXT_API_ENABLED
        -Dspm_CONFIG_SPM_SERVICE_PREVALIDATE

tejlmand replied:
> But I get this warning (when running through sanitycheck, not when doing it manually with west build)

that worries me a bit.
sanitycheck should not get other warnings/errors than what is seen on command line.
How did you invoke `sanitycheck` ?

oyvindronningstad replied:
> How did you invoke `sanitycheck` ?

Try:

    ../zephyr/scripts/sanitycheck -v -N -T samples/nrf9160/secure_services/ --platform nrf9160dk_nrf9160ns
    
    cat sanity-out/nrf9160dk_nrf9160ns/samples.nrf9160.secure_services.bootloaders/build.log | grep Manually-specified

oyvindronningstad replied:
@tejlmand is there anything we can do with this?

oyvindronningstad replied:
Finally figured it out, silly mistake. `extra_args` shouldn't have the `-D` prepended.

oyvindronningstad commented:
> It looks good to me but I don't really get why we need this?
> Is it for when you use the sample but have disabled certain `SPM_SERVICES`?

Yes, the RNG service is not available on nRF53, and also takes significant flash space. The prevalidate service only makes sense if the immutable bootloader is included.

> Do you think we could use `shared_vars.cmake` to propagate some information from SPM instead? As this will still allow you to set invalid configurations right?

That's possible, but involves some manual Cmake code to propagate the config into C code. I think it will be more non-standard and less transparent than the current solution, but if this is preferred, I will look into it.

sigvartmh commented:
It's fine by me if you want to keep the current solution I don't see it leaking too much into the rest so I would say for now the solution is good enough. When we get more services we may opt-in to look a bit further into how to improve it.

hakonfam commented on samples/nrf9160/secure_services/Kconfig:
```diff
@@ -0,0 +1,30 @@
+
+# NORDIC_SECURITY_BACKEND is not supported on 53 yet
+	depends on ! SOC_NRF5340_CPUAPP
+
```
```suggestion
```

oyvindronningstad replied:
This is copied from subsys/spm/Kconfig. It's there to make it noticeable, I think, since we want to remove it eventually.

hakonfam commented:
@oyvindronningstad CI is having hickups, this should be fixed with a rebase.

tejlmand commented on samples/nrf9160/secure_services/Kconfig:
```diff
@@ -0,0 +1,30 @@
+	bool "Enable code that calls the spm_request_random_number()"
+	default y
+
+# NORDIC_SECURITY_BACKEND is not supported on 53 yet
```
you mean the `NORDIC_SECURITY_BACKEND` or the cc310 ?

To my knowledge, you can still use `NORDIC_SECURITY_BACKEND` on nRF53 with mbed TLS vanilla backend.

oyvindronningstad replied:
This is copied from subsys/spm/Kconfig. @frkv, is this restriction lifted now? I tried building on nrf53 and it did not seem to build.

oyvindronningstad replied:
in function `spm_request_random_number':
    /home/oyro/git/ncs/nrf/subsys/spm/secure_services.c:121: undefined reference to `mbedtls_hardware_poll'

tejlmand replied:
seems wrong to have that test, would `depends on CC310_ENTROPY` be more correct ?
as that seems to be the foundation for providing `mbedtls_hardware_poll` funtionality.

https://github.com/nrfconnect/sdk-nrfxlib/blob/master/nrf_security/src/mbedtls/CMakeLists.txt#L134-L138

@frkv Maybe nRF Security should provide a config setting that indicates whether `mbedtls_hardware_poll` is available or not.
Direct checks for `! SOC_NRF5340_CPUAPP` seems wrong to me.

frkv replied:
Yes. This should point to the CC310 entropy driver, which has a config. In time it will point to CC312 entropy driver as well, but this would need to go in in a different PR

It won't build for nRF5340 ATM.

tejlmand commented on samples/nrf9160/secure_services/Kconfig:
```diff
@@ -0,0 +1,30 @@
+	default y
+
+# NORDIC_SECURITY_BACKEND is not supported on 53 yet
+	depends on ! SOC_NRF5340_CPUAPP
```
seems wrong to have this dependency, is there nothing else you can depend upon, like `NORDIC_SECURITY_BACKEND` or `cc3xx` or ?
So that when it is supported on nRF53, then this code will automatically be active.

oyvindronningstad replied:
Makes sense. @tejlmand or @frkv please tell me if there is a symbol this can depend on that is better than SOC_NRF5340_CPUAPP.

Alternatively, we can change it in a separate PR.

frkv replied:
Yes. This should point to the CC310 entropy driver, which has a config. In time it will point to CC312 entropy driver as well, but this would need to go in in a different PR

It won't build for nRF5340 ATM.

hakonfam reviewed (changes requested):
You need NCSDK- tags on the last two commits.

Typo in last commit message "pverflowed"

Is the way you set the `MAIN_STACK_SIZE` really how it should be done?

oyvindronningstad commented:
> You need NCSDK- tags on the last two commits.
> 
> Typo in last commit message "pverflowed"

Thanks, fixed.
 
> Is the way you set the `MAIN_STACK_SIZE` really how it should be done?

Thanks, I changed it to be specified in spm's prj.conf instead, I think that's better.

oyvindronningstad commented:
@hakonfam merge? :)

hakonfam commented:
@oyvindronningstad I have lost my merge rights, not sure if I want them back :D 

@tejlmand @ioannisg could one of you merge this?

oyvindronningstad commented:
@ioannisg merge?
